router/src/main/java/org/simple/clinic/router/screen/FullScreenKey.kt

```diff
@@ -16,4 +19,66 @@ interface FullScreenKey : Parcelable {
 
   @LayoutRes
   fun layoutRes(): Int
+
```

Comment:
Can we refactor this so that the responsibility of running the animations remains with the `FullScreenKeyChanger`, which delegates instantiation of the animations themselves to the key?

I tried this out and it seems to be a cleaner approach:
### `FullScreenKey`
```kotlin
  companion object {
    private const val scaleChange = 0.05f
    private const val duration = SCREEN_CHANGE_ANIMATION_DURATION

    // Interpolators are thread-safe, so we can share an instance without problems.
    private val interpolator by lazy(LazyThreadSafetyMode.NONE) { FastOutSlowInInterpolator() }
  }

  @LayoutRes
  fun layoutRes(): Int

  fun incomingViewForwardAnimation(view: View): ViewPropertyAnimator {
    view.scaleX = 1f + scaleChange
    view.scaleY = 1f + scaleChange
    view.alpha = 0f

    return view.animate()
        .scaleX(1f)
        .scaleY(1f)
        .alpha(1f)
        .setDuration(duration)
        .setInterpolator(interpolator)
  }

  fun outgoingViewForwardAnimation(view: View): ViewPropertyAnimator {
    return view.animate()
        .scaleX(1f - scaleChange)
        .scaleY(1f - scaleChange)
        .alpha(0f)
        .setDuration(duration)
        .setInterpolator(interpolator)
  }

  fun incomingViewAnimation(view: View): ViewPropertyAnimator {
    view.scaleX = 1f - scaleChange
    view.scaleY = 1f - scaleChange
    view.alpha = 0f

    return view.animate()
        .scaleX(1f)
        .scaleY(1f)
        .alpha(1f)
        .setDuration(duration)
        .setInterpolator(interpolator)
  }

  fun outgoingViewAnimation(view: View): ViewPropertyAnimator {
    return view.animate()
        .scaleX(1f + scaleChange)
        .scaleY(1f + scaleChange)
        .alpha(0f)
        .setDuration(duration)
        .setInterpolator(interpolator)
  }
```

### `FullScreenKeyChanger`
```kotlin
override fun animate(
      outgoingKey: FullScreenKey?,
      outgoingView: View?,
      incomingKey: FullScreenKey,
      incomingView: View,
      direction: flow.Direction,
      onCompleteListener: () -> Unit
  ) {
    if (outgoingView == null || direction == Direction.REPLACE) {
      onCompleteListener.invoke()
      return
    }

    val outGoingViewAnimation = when (direction) {
      Direction.FORWARD -> incomingKey.outgoingViewForwardAnimation(outgoingView)
      else -> incomingKey.outgoingViewAnimation(outgoingView)
    }

    val incomingViewAnimation = when (direction) {
      Direction.FORWARD -> incomingKey.incomingViewForwardAnimation(incomingView)
      else -> incomingKey.incomingViewAnimation(incomingView)
    }

    outGoingViewAnimation.start()
    incomingViewAnimation
        .withEndAction { onCompleteListener.invoke() }
        .start()
  }
```

Reply:
This gives us a couple of advantages:
- Whenever we override an animation, we will only need to replace the one we care about.
- In any `FullScreenKey` where we override the animation, we will no longer have to remember to run the completion listener since `FullScreenKeyChanger` will do that for us.

Reply:
Yep, this was the idea Rakshak and I discussed. @rakshakhegde do you want to go ahead and make this change?

Reply:
As we have reverted back to Activity, this animation for Flow views should no longer be part of this PR. So I have deleted the commit with the animation. Let's pick this up in a later PR. What do y'all think?

Reply:
Makes sense